metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala

```diff
@@ -117,89 +113,126 @@ object FileWatcher {
+    val stopWatchingSignal: AtomicBoolean = new AtomicBoolean
+
+    // queue is processed serially to defensively prevent possible races
+    // when a file is created and then deleted in quick succesion.
+    // Upstream callbacks implicitly assume that events for a single file
+    // arrive and are processed serially.
+    // This consumption could be parallelized, but in case in indexing,
+    // compilation is usually the dominating bottleneck
+    val thread = new Thread("metals-watch-callback-thread") {
+      @tailrec def loop(): Unit = {
+        try callback(watchEventQueue.take)
+        catch { case _: InterruptedException => }
+        if (!stopWatchingSignal.get) loop()
+      }
+      override def run(): Unit = loop()
+      start()
+    }
 
-      () => repo.close()
+    () => {
+      watcher.close()
+      stopWatchingSignal.set(true)
+      thread.interrupt()
     }
   }
 
-  private def initFileTreeRepository(
+  /**
+   * Initialize file watcher
+   *
+   * File watch events are put into a queue which is processed separately
+   * to prevent callbacks to be executed on the thread that watches
+   * for file events.
+   *
+   * @param watchFilter for incoming file watch events
+   * @param queue for file events
+   * @return
+   */
+  private def initWatcher(
       watchFilter: Path => Boolean,
-      callback: FileWatcherEvent => Unit
-  ): FileTreeRepository[Hash] = {
-    val converter: Converter[Hash] = typedPath =>
-      hashFile(
-        typedPath.getPath(),
-        watchFilter
-      )
-    val repo = FileTreeRepositories.get(converter, /*follow symlinks*/ true)
-
-    repo.addCacheObserver(new CacheObserver[Hash] {
-      override def onCreate(entry: Entry[Hash]): Unit = {
-        val path = entry.getTypedPath().getPath()
-        if (watchFilter(path)) callback(FileWatcherEvent.create(path))
-      }
-      override def onDelete(entry: Entry[Hash]): Unit = {
-        val path = entry.getTypedPath().getPath()
-        if (watchFilter(path)) callback(FileWatcherEvent.delete(path))
+      queue: BlockingQueue[FileWatcherEvent]
+  ): PathWatcher[PathWatchers.Event] = {
+    val watcher = PathWatchers.get( /*follow symlinks*/ true)
+
+    watcher.addObserver(new Observer[PathWatchers.Event] {
+      override def onError(t: Throwable): Unit = {
+        scribe.error(s"Error encountered during file watching: ${t}")
       }
-      override def onUpdate(
-          previous: Entry[Hash],
-          current: Entry[Hash]
-      ): Unit = {
-        val path = current.getTypedPath().getPath()
-        if (previous.getValue != current.getValue && watchFilter(path)) {
-          callback(FileWatcherEvent.modify(path))
+
+      override def onNext(event: PathWatchers.Event): Unit = {
+        val path = event.getTypedPath.getPath
+        if (watchFilter(path)) {
+          event.getKind match {
+            case Kind.Create => {
+              queue.add(FileWatcherEvent.create(path))
+            }
```

Comment:
Could you remove the brackets surrounding the cases' bodies?

Reply:
> Looks good, I have only a few suggestions. Also, I'm going to test this change manually today.

Thanks! I will address your comments this evening. I also noticed that I forgot to address one issue, see https://github.com/scalameta/metals/pull/3758#discussion_r841469303